Add docs for placement rules in SQL (#6618)
pingcap/docs: placement-rules-in-sql.md

---
title: Placement Rules in SQL
summary: Learn how to schedule placement of tables and partitions.
---

# Placement Rules in SQL

> **Warning:**
>
> Placement Rules in SQL is an experimental feature. The syntax might change before its GA, and there might also be bugs.
>
> If you understand the risks, you can enable this experiment feature by executing `SET GLOBAL tidb_enable_alter_placement = 1;`.

> Placement Rules allow you to configure where data will be stored in a TiKV cluster. This is useful for scenarios including optimizing a high availability strategy, ensuring that local copies of data will be available for local stale reads, and adhering to compliance requirements.

Reviewer: Could you explain what is Placement Rules? What is the relationship between the existing [Placement Rules implemented via PD](https://docs.pingcap.com/tidb/dev/configure-placement-rules)?

Maybe some text in https://docs.pingcap.com/tidb/dev/configure-placement-rules explains in more detail. Is is possible to borrow or copy some of the text from that document? For example:

It is a replica rule system that guides PD to generate corresponding schedules for different types of data. By combining different scheduling rules, you can finely control the attributes of any continuous data range, such as the number of replicas, the storage location, the host type, whether to participate in Raft election, and whether to act as the Raft leader.

Author: I actually think it does state what they do this in this sentence:

> Placement Rules allow you to configure where data will be stored in a TiKV cluster. This is useful for scenarios including optimizing a high availability strategy, ensuring that local copies of data will be available for local stale reads, and adhering to compliance requirements.

The only part that it misses is the relationship to the previous implementation which is it is _configurable in SQL._

This part is too complicated for users:

> It is a replica rule system that guides PD to generate corresponding schedules for different types of data. By combining different scheduling rules, you can finely control the attributes of any continuous data range, such as the number of replicas, the storage location, the host type, whether to participate in Raft election, and whether to act as the Raft leader.

It is actually not our goal to have multiple different schedules combined, since that increases the risk of misconfiguration. This is why we created the concept of a `PLACEMENT POLICY`, so you can state one policy and apply it everywhere.

Reviewer: Thanks for explaning. LGTM.

> ## Specifying placement options

Reviewer: ```suggestion
## Specify placement options
```

Placement options can be specified using either _direct placement_ or using a _placement policy_. In the following example, both tables `t1` and `t2` have the same rules:

```sql
CREATE TABLE t1 (a INT) PRIMARY_REGION="us-east-1" REGIONS="us-east-1,us-west-1";
CREATE PLACEMENT POLICY eastandwest PRIMARY_REGION="us-east-1" REGIONS="us-east-1,us-west-1";
CREATE TABLE t2 (a INT) PLACEMENT POLICY=eastandwest;
```

> Using placement policies is recommended to simplify rule management. Changes made to a placement policy (via [`ALTER PLACEMENT POLICY`](/sql-statements/sql-statement-alter-placement-policy.md)) automatically propagate to all database objects. This differs from direct placement options, which the rules must be altered for each object (tables, partitions).

Reviewer: ```suggestion
It is recommended to use placement policies for simpler rule management. When you change a placement policy (via [`ALTER PLACEMENT POLICY`](/sql-statements/sql-statement-alter-placement-policy.md)), the change automatically propagates to all database objects.

If you use direct placement options, you have to alter rules for each object (for example, tables and partitions).
```

> A `PLACEMENT POLICY` is not associated with any database schema and has global scope. Assigning a placement policy does not require any additional privileges over `CREATE TABLE` privilege.

Reviewer: ```suggestion
`PLACEMENT POLICY` is not associated with any database schema and has the global scope. Therefore, assigning a placement policy does not require any additional privileges over the `CREATE TABLE` privilege.
```

## Option reference

> **Note:**
>

> > Placement options depend on labels correctly specified in the configuration of each TiKV node. For example, `PRIMARY_REGION` depends on the label `region` in TiKV. To see a summary of all labels available in your TiKV cluster, use the statement [`SHOW PLACEMENT LABELS`](/sql-statements/sql-statement-show-placement-labels.md):

Reviewer: ```suggestion
> Placement options depend on labels correctly specified in the configuration of each TiKV node. For example, the `PRIMARY_REGION` option depends on the `region` label in TiKV. To see a summary of all labels available in your TiKV cluster, use the statement [`SHOW PLACEMENT LABELS`](/sql-statements/sql-statement-show-placement-labels.md):
```

>
> ```sql
> mysql> show placement labels;
> +--------+----------------+
> | Key    | Values         |
> +--------+----------------+
> | disk   | ["ssd"]        |
> | region | ["us-east-1"]  |
> | zone   | ["us-east-1a"] |
> +--------+----------------+
> 3 rows in set (0.00 sec)
> ```

| Option Name                | Description                                                                                    |
|----------------------------|------------------------------------------------------------------------------------------------|

> | `PRIMARY_REGION`           | Raft leaders will be placed in stores which have the `region` label that match this value.     |

Reviewer: Does "this value" mean the `PRIMARY_REGION` value?
```suggestion
| `PRIMARY_REGION`           | Raft leaders are placed in stores which have the `region` label that matches this value.     |
```

| `REGIONS`                  | Raft followers will be placed in stores which have the `region` label that matches this value. |

> | `SCHEDULE`                 | The strategy used to schedule the placement of followers. Either `EVEN` (default) or `MAJORITY_IN_PRIMARY`. |

Reviewer: ```suggestion
| `SCHEDULE`                 | The strategy used to schedule the placement of followers. The value options are `EVEN` (default) or `MAJORITY_IN_PRIMARY`. |
```

> | `FOLLOWERS`                | The number of followers. For example, `FOLLOWERS=2` means that there will be 3 copies of the data (2 followers and 1 leader). |

Reviewer: ```suggestion
| `FOLLOWERS`                | The number of followers. For example, `FOLLOWERS=2` means that there will be 3 replicas of the data (2 followers and 1 leader). |
```

> | **Advanced configuration**                                                                                                  |
> | `CONSTRAINTS`              | A list of constraints that apply to all roles. For example, `CONSTRAINTS="[+disk=ssd]`.        |
> | `FOLLOWER_CONSTRAINTS`     | A list of constraints that only apply to followers.                                            |

Reviewer: ```suggestion
In addition to the placement options above, you can also use the advance configurations. For details, see [Advance placement](#advanced-placement).

| Option Name                | Description                                                                                    |
|----------------------------|------------------------------------------------------------------------------------------------|                                                                                       
| `CONSTRAINTS`              | A list of constraints that apply to all roles. For example, `CONSTRAINTS="[+disk=ssd]`.        |
| `FOLLOWER_CONSTRAINTS`     | A list of constraints that only apply to followers.                                            |
```

## Examples

> ### Increasing the number of replicas

Reviewer: ```suggestion
### Increase the number of replicas
```

The default configuration of [`max-replicas`](/pd-configuration-file.md#max-replicas) is `3`. To increase this for a specific set of tables, you can use a placement policy as follows:

```sql
CREATE PLACEMENT POLICY fivereplicas FOLLOWERS=4;
CREATE TABLE t1 (a INT) PLACEMENT POLICY=fivereplicas;
```

> Note that the PD configuration includes the leader and follower count, thus 4 followers + 1 leader equals five replicas in total.

Reviewer: ```suggestion
Note that the PD configuration includes the leader and follower count, thus 4 followers + 1 leader equals 5 replicas in total.
```

> To expand on this example, the placement for the followers can also be described using the `PRIMARY_REGION` and `REGIONS` placement options:

Reviewer: ```suggestion
To expand on this example, you can also use `PRIMARY_REGION` and `REGIONS` placement options to describe the placement for the followers:
```

```sql
CREATE PLACEMENT POLICY eastandwest PRIMARY_REGION="us-east-1" REGIONS="us-east-1,us-east-2,us-west-1" SCHEDULE="MAJORITY_IN_PRIMARY" FOLLOWERS=4;
CREATE TABLE t1 (a INT) PLACEMENT POLICY=eastandwest;
```

> The `SCHEDULE` instructs TiDB how to balance the followers. The default schedule of `EVEN` ensures a balance of followers in all regions. The schedule `MAJORITY_IN_PRIMARY` can be used to ensure that enough followers are placed in the primary region (`us-east-1`) that quorum can be achieved. This helps provide lower latency transactions at the expense of availability should the primary region completely fail.

Reviewer: Does the `region` here mean data center, instead of the "Region" that implies a data range in TiKV?
```suggestion
The `SCHEDULE` option instructs TiDB on how to balance the followers.

The default schedule of `EVEN` ensures a balance of followers in all regions. You can use the `MAJORITY_IN_PRIMARY` schedule to ensure that enough followers are placed in the primary region (`us-east-1`) so that quorum can be achieved. If the primary region completely fails, you can the `MAJORITY_IN_PRIMARY` schedule for lower latency transactions at the expense of availability.
```

> ### Assigning placement to a partitioned table

Reviewer: ```suggestion
### Assign placement to a partitioned table
```

> **Note:**
>

> > This example makes use of list partitioning, which is currently an experimental feature. Partitioned tables also require the `PRIMARY KEY` to be included in all columns in the table's partitioning function.

Reviewer: ```suggestion
> The following example uses list partitioning, which is currently an experimental feature of TiDB. Partitioned tables also require the `PRIMARY KEY` to be included in all columns in the table's partitioning function.
```

> As well as assigning to tables, placement options can also be assigned to table partitions. For example:

Reviewer: ```suggestion
In addition to assigning placement options to tables, you can also assign the options to table partitions. For example:
```

```sql
CREATE PLACEMENT POLICY europe PRIMARY_REGION="eu-central-1" REGIONS="eu-central-1,eu-west-1";
CREATE PLACEMENT POLICY northamerica PRIMARY_REGION="us-east-1" REGIONS="us-east-1";

SET tidb_enable_list_partition = 1;
CREATE TABLE t1 (
  country VARCHAR(10) NOT NULL,
  userdata VARCHAR(100) NOT NULL
) PARTITION BY LIST COLUMNS (country) (
  PARTITION pEurope VALUES IN ('DE', 'FR', 'GB') PLACEMENT POLICY=europe,
  PARTITION pNorthAmerica VALUES IN ('US', 'CA', 'MX') PLACEMENT POLICY=northamerica
);
```

> ### Setting the default placement for a schema

Reviewer: ```suggestion
### Set the default placement for a schema
```

> Default placement options can be directly attached to a database schema. This works similar to setting the default character set or collation for a schema, in that it will be used when no other placement options are specified. For example:

Reviewer: ```suggestion
You can directly attach the default placement options to a database schema. This works similar to setting the default character set or collation for a schema. Your specified placement options apply when no other options are specified. For example:
```

```sql

> CREATE TABLE t1 (a INT); -- the table is created with no placement options
> ALTER DATABASE test FOLLOWERS=4; -- this changes the default, and does not apply to the existing table t1;
> CREATE TABLE t2 (a INT); -- the placement of FOLLOWERS=4 will be used
> CREATE TABLE t3 (a INT) PRIMARY_REGION="us-east-1" REGIONS="us-east-1,us-east-2"; -- FOLLOWERS=4 does not apply as placement is specified.
> ALTER DATABASE test FOLLOWERS=2; -- this does not apply to existing tables
> CREATE TABLE t4 (a INT); -- the table is created with FOLLOWERS=2;

Reviewer: ```suggestion
CREATE TABLE t1 (a INT);  -- Creates a table t1 with no placement options.

ALTER DATABASE test FOLLOWERS=4;  -- Changes the default placement option, and does not apply to the existing table t1.

CREATE TABLE t2 (a INT);  -- Creates a table t2 with the default placement of FOLLOWERS=4.

CREATE TABLE t3 (a INT) PRIMARY_REGION="us-east-1" REGIONS="us-east-1,us-east-2";  -- Creates a table t3 without the default FOLLOWERS=4 placement, because this statement has specified another placement.

ALTER DATABASE test FOLLOWERS=2;  -- Changes the default placement, and does not apply to existing tables.

CREATE TABLE t4 (a INT);  -- Creates a table t4 with the default FOLLOWERS=2 option.
```

```

> Because placement options are only inherited from the database schema default when a table is created, it is recommended to set the default to a `PLACEMENT POLICY`. This ensures that future changes to the policy will propagate to existing tables.

Reviewer: ```suggestion
Because placement options are only inherited from the database schema when a table is created, it is recommended to set the default placement option using a `PLACEMENT POLICY`. This ensures that future changes to the policy propagate to existing tables.
```

);
```

Constraints can either be specified in list format (`[+disk=ssd]`) or dictionary format (`{+disk=ssd:1,+disk=hdd:2}`).

In list format, constraints are specified as a list of key-value pairs. The key starts with either a `+` or a `-` with `+disk=ssd` indicating that the label `disk` must be set to `ssd`, and `-disk=hdd` indicating that the label `disk` must not be `hdd`.

In dictionary format, the constraint also indicates a number of instances that apply to that rule. For example `FOLLOWER_CONSTRAINTS="{+region=us-east-1:1,+region=us-east-2:1,+region=us-west-1:1,+any:1}";` indicates that 1 follower is in us-east-1, 1 follower is in us-east-2, 1 follower is in us-west-1, and 1 follower can be in any region.

## Known limitations

The following known limitations exist in the experimental release of Placement Rules in SQL:

* Dumpling does not support dumping placement policies. See [issue #29371](https://github.com/pingcap/tidb/issues/29371)
* TiDB tools including BR, CDC, Lightning and DM do not yet support placement rules.
* Temporary tables do not support placement options (either via direct placement or placement policies).

> * Syntactic sugar rules exist for setting `PRIMARY_REGION` and `REGIONS`, but in future we plan to add varieties for `PRIMARY_RACK`, `PRIMARY_ZONE` and `PRIMARY_HOST`. See [issue #18030](https://github.com/pingcap/tidb/issues/18030)

Author: Yes, but this is under the headline "Known limitations" as it applies to the experimental release. This is a known limitation, since there are other configurations besides regionality.

> * Syntactic sugar rules exist for setting `PRIMARY_REGION` and `REGIONS`, but in future we plan to add varieties for `PRIMARY_RACK`, `PRIMARY_ZONE` and `PRIMARY_HOST`. See [issue #18030](https://github.com/pingcap/tidb/issues/18030)

Reviewer: Get it. I've modified my comment. PTAL

`PLACEMENT POLICY` is not associated with any database schema and has the global scope. Therefore, assigning a placement policy does not require any additional privileges over the `CREATE TABLE` privilege.

## Option reference

> **Note:**
>
> Placement options depend on labels correctly specified in the configuration of each TiKV node. For example, the `PRIMARY_REGION` option depends on the `region` label in TiKV. To see a summary of all labels available in your TiKV cluster, use the statement [`SHOW PLACEMENT LABELS`](/sql-statements/sql-statement-show-placement-labels.md):
>
> ```sql
> mysql> show placement labels;
> +--------+----------------+
> | Key    | Values         |
> +--------+----------------+
> | disk   | ["ssd"]        |
> | region | ["us-east-1"]  |
> | zone   | ["us-east-1a"] |
> +--------+----------------+
> 3 rows in set (0.00 sec)
> ```

| Option Name                | Description                                                                                    |
|----------------------------|------------------------------------------------------------------------------------------------|

> | `PRIMARY_REGION`           | Raft leaders are placed in stores which have the `region` label that matches this value.     |
> | `REGIONS`                  | Raft followers will be placed in stores which have the `region` label that matches this value. |
> | `SCHEDULE`                 | The strategy used to schedule the placement of followers. The value options are `EVEN` (default) or `MAJORITY_IN_PRIMARY`. |
> | `FOLLOWERS`                | The number of followers. For example, `FOLLOWERS=2` means that there will be 3 replicas of the data (2 followers and 1 leader). |
> In addition to the placement options above, you can also use the advance configurations. For details, see [Advance placement](#advanced-placement).

Reviewer: ```suggestion
| `PRIMARY_REGION`           | Raft leaders are placed in stores that have the `region` label that matches the value of this option.     |
| `REGIONS`                  | Raft followers are placed in stores that have the `region` label that matches the value of this option. |
| `SCHEDULE`                 | The strategy used to schedule the placement of followers. The value options are `EVEN` (default) or `MAJORITY_IN_PRIMARY`. |
| `FOLLOWERS`                | The number of followers. For example, `FOLLOWERS=2` means that there will be 3 replicas of the data (2 followers and 1 leader). |

In addition to the placement options above, you can also use the advance configurations. For details, see [Advance placement](#advanced-placement).
```

---
title: Placement Rules in SQL
summary: Learn how to schedule placement of tables and partitions.
---

# Placement Rules in SQL

> **Warning:**
>
> Placement Rules in SQL is an experimental feature. The syntax might change before its GA, and there might also be bugs.
>

> > If you understand the risks, you can enable this experiment feature by executing `SET GLOBAL tidb_enable_alter_placement = 1;`.

Reviewer 2: Can we add a note like:

```
We recommend using this feature with the help of the official tidb team to help solve specific scenarios. Over-configuration of the placement plan will increase the difficulty of maintenance and troubleshooting, so reasonable planning is very necessary. Usually, the points that need to be considered, including disaster recovery issues, high availability issues, performance issues and so on.
```

Author: Hi @nolouch. Some comments in line:

> We recommend using this feature with the help of the official tidb team to help solve specific scenarios. Over-configuration of the placement plan will increase the difficulty of maintenance and troubleshooting, so reasonable planning is very necessary.

To a large degree we removed over configuration as a possibility between the original proposal and the newer proposal. You can no longer specify multiple constraints; and syntactic sugar constraints are mutually exclusive with regular constraints. The only complex scenarios that remain are both `FOLLOWER_CONSTRAINTS` and `CONSTRAINTS` can apply, and a user could set a complex constraint in dictionary format.

>  Usually, the points that need to be considered, including disaster recovery issues, high availability issues, performance issues and so on.

We don't document this well in the current experimental version. Because it is experimental, users are not encouraged to use it in production, so high availability/DR did not seem relevant (yet).

I have an alternative suggestion:

> We value your feedback when evaluating experimental features. Please feel free to [create a GitHub issue](https://github.com/pingcap/tidb/issues/new/choose) or reach out to us on [TiDB Slack](https://slack.tidb.io/invite?team=tidb-community&channel=everyone&ref=pingcap). We are interested to both the problems you encounter and your intended use-cases.
